ccip.go

```diff
@@ -0,0 +1,163 @@
+package ens
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/wealdtech/go-ens/v3/contracts/offchainresolver"
+	"github.com/wealdtech/go-ens/v3/contracts/universalresolver"
+)
+
+func getCcipReadError(err error) (bool, string) {
+	var jsonErr, ok = err.(rpc.DataError)
+	if !ok {
+		return false, ""
+	}
+	errData, ok := jsonErr.ErrorData().(string)
+	if !ok {
+		return false, ""
+	}
+	return (len(errData) >= 10 && errData[:10] == offchainLookupSignature), errData
+}
+
+func CCIPRead(backend bind.ContractBackend, rAddr common.Address, revertData string) ([]byte, error) {
+	hexBytes, err := hex.DecodeString(revertData[2:])
+	if err != nil {
+		return nil, err
+	}
+
+	uAbi, err := universalresolver.ContractMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
+	// Extracting error details from the revert data
+	var sig [4]byte
+	copy(sig[:], hexBytes[:4])
+	abiErr, err := uAbi.ErrorByID(sig)
+	if err != nil {
+		return nil, err
+	}
+	errArgs, err := abiErr.Inputs.Unpack(hexBytes[4:])
+	if err != nil {
+		return nil, err
+	}
+
+	sender := errArgs[0].(common.Address)
```

Comment:
There is a lot of casting and array access here without checks, please add checks to avoid potential panics.

Reply:
for the returned error to have the `OffchainLookup` signature (`0x556f1830`), it must have the following arguments:

* sender
* urls
* calldata
* callback
* extraData

so I assume those variables will be on the `errArgs` variable and so they can be accessed directly by the index. otherwise, it would have failed on the `(len(errData) >= 10 && errData[:10] == offchainLookupSignature)` [condition](https://github.com/wealdtech/go-ens/pull/41/files#diff-909eb35e6fc9e86c3d9d84853cc960d3b2ef2eedfdb49e099ccca547f8e98550R31)

Reply:
It is easy for an error to be returned with the same four bytes but representing a different function (see [4byte.info](https://www.4byte.directory/signatures/?bytes4_signature=0x70a08231) for an example of clashes), and if done maliciously then it would cause a panic.